### PR TITLE
Split handler context

### PR DESCRIPTION
**Pull request by timbussmann**

### CODE REVIEW

**Comment by timbussmann** [comment]
@andreasohlund @danielmarbach what do you think about these changes?
Please note the changes in the `ContextualBus`. A further step could be to remove the whole logic besides getting the current context into a `StaticBus` (better naming required) which always takes a context as a parameter for all operations. As long as we support DI of some `IBus` interface anywhere were unable to remove the `ContextStacker` & `ContextualBus` itself.

**Comment by andreasohlund** [comment]
>As long as we support DI of some IBus interface anywhere were unable to remove the ContextStacker & ContextualBus itself.

Wasn't the plan to not inject either IBus(will be deprecated) or ISendOnlyBus (availiable via Endpoint.Start())?

**Comment by andreasohlund** [comment]
I like the changes, :+1:

**Comment by danielmarbach** [comment]
Remove the Async postfix?

On Friday, October 23, 2015, Tim Bussmann <notifications@github.com> wrote:

> @andreasohlund <https://github.com/andreasohlund> @danielmarbach
> <https://github.com/danielmarbach> what do you think about these changes?
> Please note the changes in the ContextualBus. A further step could be to
> remove the whole logic besides getting the current context into a
> StaticBus (better naming required) which always takes a context as a
> parameter for all operations. As long as we support DI of some IBus
> interface anywhere were unable to remove the ContextStacker &
> ContextualBus itself.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/3030#issuecomment-150622673>
> .
>


-- 
*#NSBcon 2015 • The "All About NServiceBus" conference is back *
*• Learn more <http://go.particular.net/l/27442/2015-08-21/5d4wzp> →*

**Comment by danielmarbach** [comment]
Yes that was the plan

On Monday, October 26, 2015, Andreas Öhlund <notifications@github.com>
wrote:

> As long as we support DI of some IBus interface anywhere were unable to
> remove the ContextStacker & ContextualBus itself.
>
> Wasn't the plan to not inject either IBus(will be deprecated) or
> ISendOnlyBus (availiable via Endpoint.Start())?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/3030#issuecomment-151046226>
> .
>


-- 
*#NSBcon 2015 • The "All About NServiceBus" conference is back *
*• Learn more <http://go.particular.net/l/27442/2015-08-21/5d4wzp> →*

**Comment by timbussmann** [comment]
@andreasohlund 
> Wasn't the plan to not inject either IBus(will be deprecated) or ISendOnlyBus (availiable via Endpoint.Start())?

That's what I understood as long-term goal. We need to write those decisions down after a meeting :( As long there isn't a context available somehow everywhere a user can extend NSB (which is basically everywhere?), how should users migrate to use the contextual alternatives?

@danielmarbach we should do that as a separate PR.

**Comment by danielmarbach** [comment]
For some "extension points" a factory to create a SendContext or a
SendContext provided by the core should be sufficient

On Monday, October 26, 2015, Tim Bussmann <notifications@github.com> wrote:

> @andreasohlund <https://github.com/andreasohlund>
>
> Wasn't the plan to not inject either IBus(will be deprecated) or
> ISendOnlyBus (availiable via Endpoint.Start())?
>
> That's what I understood as long-term goal. We need to write those
> decisions down after a meeting :( As long there isn't a context available
> somehow everywhere a user can extend NSB (which is basically everywhere?),
> how should users migrate to use the contextual alternatives?
>
> @danielmarbach <https://github.com/danielmarbach> we should do that as a
> separate PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/3030#issuecomment-151051656>
> .
>


-- 
*#NSBcon 2015 • The "All About NServiceBus" conference is back *
*• Learn more <http://go.particular.net/l/27442/2015-08-21/5d4wzp> →*

**Comment by timbussmann** [comment]
@danielmarbach Currently, the contexts we have are basically 1:1 the same thing as the behavior context used in the pipeline. What would such a factory do in order to access the current behavior context? (that's basically the thing `ContextualBus` and `ContextStacker` do?)

File: src/NServiceBus.Core/Unicast/ContextualBus.cs [hunk 1]
```diff
@@ -101,43 +102,48 @@ public Task ReplyAsync(object message, NServiceBus.ReplyOptions options)
...
-            return ReplyAsync(messageMapper.CreateInstance(messageConstructor), options);
+            return ReplyAsync(messageMapper.CreateInstance(messageConstructor), options, incomingContext);
         }
 
         /// <summary>
         /// Moves the message being handled to the back of the list of available 
         /// messages so it can be handled later.
         /// </summary>
-        public async Task HandleCurrentMessageLaterAsync()
+        public async Task HandleCurrentMessageLaterAsync(InvokeHandlerContext incomingContext)
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Unicast/ContextualBus.cs]: rename to just context?

File: src/NServiceBus.Core/Unicast/ContextualBus.cs [hunk 2]
```diff
@@ -101,43 +102,48 @@ public Task ReplyAsync(object message, NServiceBus.ReplyOptions options)
...
         {
+            var messageBeingProcessed = incomingContext.Get<IncomingMessage>();
+
             var pipeline = new PipelineBase<RoutingContext>(builder, settings, settings.Get<PipelineConfiguration>().MainPipeline);
 
-            var outgoingMessage = new OutgoingMessage(MessageBeingProcessed.MessageId, MessageBeingProcessed.Headers, MessageBeingProcessed.Body);
+            var outgoingMessage = new OutgoingMessage(messageBeingProcessed.MessageId, messageBeingProcessed.Headers, messageBeingProcessed.Body);
             var context = new RoutingContext(outgoingMessage, new UnicastRoutingStrategy(destination), incomingContext);
 
             await pipeline.Invoke(context);
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Unicast/ContextualBus.cs]: Seems we missed a `ConfigureAwait(false)`

File: src/NServiceBus.Core/Unicast/ContextualBus.cs (hunk 1 above)
Comment (timbussmann) [inline src/NServiceBus.Core/Unicast/ContextualBus.cs]: I wanted to extract these methods to a new class anyway as a next step. Let's rename it there?

**Comment by timbussmann** [comment]
@danielmarbach fixed your comments

**Pull request merged**